airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/StateDecoratingIterator.java

```diff
@@ -66,42 +73,98 @@ private String getCursorCandidate(final AirbyteMessage message) {
     return (cursorCandidate != null ? cursorCandidate.replaceAll("\u0000", "") : null);
   }
 
+  /**
+   * Computes the next record retrieved from Source stream. Emits StateMessage containing data of the
+   * record that has been read so far
+   *
+   * <p>
+   * If this method throws an exception, it will propagate outward to the {@code hasNext} or
+   * {@code next} invocation that invoked this method. Any further attempts to use the iterator will
+   * result in an {@link IllegalStateException}.
+   * </p>
+   *
+   * @return {@link AirbyteStateMessage} containing information of the records read so far
+   */
   @Override
   protected AirbyteMessage computeNext() {
+    if (hasCaughtException) {
+      // Mark iterator as done since the next call to messageIterator will result in an
+      // IllegalArgumentException and resets exception caught state.
+      // This occurs when the previous iteration emitted state so this iteration cycle will indicate
+      // iteration is complete
+      hasCaughtException = false;
+      return endOfData();
+    }
+
     if (messageIterator.hasNext()) {
-      if (emitIntermediateState && intermediateStateMessage != null) {
-        final AirbyteMessage message = intermediateStateMessage;
-        intermediateStateMessage = null;
-        emitIntermediateState = false;
-        return message;
+      Optional<AirbyteMessage> optionalIntermediateMessage = getIntermediateMessage();
+      if (optionalIntermediateMessage.isPresent()) {
+        return optionalIntermediateMessage.get();
       }
 
       totalRecordCount++;
-      final AirbyteMessage message = messageIterator.next();
-      if (message.getRecord().getData().hasNonNull(cursorField)) {
-        final String cursorCandidate = getCursorCandidate(message);
-        if (IncrementalUtils.compareCursors(maxCursor, cursorCandidate, cursorType) < 0) {
-          if (stateEmissionFrequency > 0 && !Objects.equals(maxCursor, initialCursor) && messageIterator.hasNext()) {
-            // Only emit an intermediate state when it is not the first or last record message,
-            // because the last state message will be taken care of in a different branch.
-            intermediateStateMessage = createStateMessage(false);
+      // Use try-catch to catch Exception that could occur when connection to the database fails
+      try {
+        final AirbyteMessage message = messageIterator.next();
+        if (message.getRecord().getData().hasNonNull(cursorField)) {
+          final String cursorCandidate = getCursorCandidate(message);
+          if (IncrementalUtils.compareCursors(maxCursor, cursorCandidate, cursorType) < 0) {
+            if (stateEmissionFrequency > 0 && !Objects.equals(maxCursor, initialCursor) && messageIterator.hasNext()) {
+              // Only emit an intermediate state when it is not the first or last record message,
```

Comment:
- why not emit intermediate state if this is the first record message?
 - where is the last state message taken care of?

Reply:
* it does not emit state since after reading the first record it is not clear if there's records with the same `cursorField`, once a cursor has been updated it will emit the state on the next iteration. This example is explained a little further in [this comment](https://github.com/airbytehq/airbyte/pull/15496#discussion_r942590918) and in `StateDecoratingIteratorTest#testStateEmissionForRecordsSharingSameCursorValue`
* the last state message is _only_ emitted when it reaches a "ready" state. In this case the last state message is taken care of within the `StateDecoratingIterator#createStateMessage`. To the best of my knowledge there is also still work that Liren is completing with regards to the last state message in [this PR](https://github.com/airbytehq/airbyte/pull/15535/files)